### PR TITLE
feat: allow filtering of validators by top-level validator status

### LINKED ISSUE #7141: Allow filtering of validators by top-level validator status
Right now, Lodestar does not support filtering response of [getStateValidators](https://ethereum.github.io/beacon-APIs/#/Beacon/getStateValidators) by top-level validator statuses such as `active` as we only consider the more accurate sub statuses like `active_ongoing`, `active_exiting` and `active_slashed`.

As per beacon-api spec we should support `active`, `pending`, `exited`, and `withdrawal` ([validators.yaml#L38](https://github.com/ethereum/beacon-APIs/blob/e7f7d70423b0abfe9d9f33b701be2ec03e44eb02/apis/beacon/states/validators.yaml#L38)).

Related code
https://github.com/ChainSafe/lodestar/blob/955f4569289d66e0a2ee171dfa7bd9984ce57ba7/packages/beacon-node/src/api/impl/beacon/state/utils.ts#L132

### PR DESCRIPTION
**Pull request by vladmarusyk**
## Add support for filtering validators by top-level status

This pull request implements support for filtering validators by top-level statuses such as `active`, `pending`, `exited`, and `withdrawal` as specified in the beacon-api spec.

### Changes:
- Added `mapToGeneralStatus` function to map detailed statuses to top-level statuses (`active`, `pending`, `exited`, and `withdrawal`).
- Updated `filterStateValidatorsByStatus` to support filtering validators by both detailed and top-level statuses.

### Related issue:
Closes #7141

### CODE REVIEW

File: packages/beacon-node/src/api/impl/beacon/state/utils.ts [hunk 1]
```diff
@@ -83,6 +83,51 @@ export async function getStateResponseWithRegen(
   return res;
 }
 
+function mapToGeneralStatus(subStatus: string): string {
```
Comment (nflaig) [inline packages/beacon-node/src/api/impl/beacon/state/utils.ts]: switch-case would be nicer here, we can also reuse `ValidatorStatus` type, and potentially a extended type which includes top-level / general statuses as well

File: packages/beacon-node/src/api/impl/beacon/state/utils.ts [hunk 2]
```diff
@@ -83,6 +83,51 @@ export async function getStateResponseWithRegen(
...
+    return "withdrawal";
+  }
+  throw new Error(`Unknown substatus: ${subStatus}`);
+}
+
+/**
+ * Get the status of the validator
+ * based on conditions outlined in https://hackmd.io/ofFJ5gOmQpu1jjHilHbdQQ
+ */
+export function getValidatorStatus(validator: phase0.Validator, currentEpoch: Epoch): routes.beacon.ValidatorStatus {
```
Comment (nflaig) [inline packages/beacon-node/src/api/impl/beacon/state/utils.ts]: why is this added here?

**Review by nflaig — COMMENTED** [review]
Thanks for picking this up, some initial feedback

**Comment by nflaig** [comment]
also note that `ValidatorStatus` type was just moved to types packages (see https://github.com/ChainSafe/lodestar/pull/7140), make sure your branch is up-to-date

**Comment by vladmarusyk** [comment]
I've updated the code based on the feedback:
- Replaced `if-else` with `switch-case`.
- Used the `ValidatorStatus` type and created an extended type for general statuses.

File: packages/beacon-node/src/api/impl/beacon/state/utils.ts [hunk 3]
```diff
@@ -83,6 +83,60 @@ export async function getStateResponseWithRegen(
   return res;
 }
 
+export type GeneralValidatorStatus = routes.beacon.ValidatorStatus | "active" | "pending" | "exited" | "withdrawal";
```
Comment (nflaig) [inline packages/beacon-node/src/api/impl/beacon/state/utils.ts]: On 2nd thought it makes more sense to just declare the top-level statuses here

```suggestion
export type GeneralValidatorStatus = "active" | "pending" | "exited" | "withdrawal";
```

**Review by nflaig — COMMENTED** [review]
Looks good, can you please remove unrelated code updates, see https://github.com/ChainSafe/lodestar/pull/7143#discussion_r1794055386

**Comment by nflaig** [comment]
https://github.com/ChainSafe/lodestar/pull/7143/commits/e07a82be26ab5c1a3258a5939dcac67d72b1be33 just removed the actual code, the unrelated code changes are still here

**Comment by vladmarusyk** [comment]
Sorry for the inconvenience, solving the first issue. Thank you for the opportunity.

**Comment by vladmarusyk** [comment]
Removed unrelated code

File: packages/beacon-node/src/api/impl/beacon/state/utils.ts [hunk 4]
```diff
@@ -83,6 +83,28 @@ export async function getStateResponseWithRegen(
   return res;
 }
 
+export type GeneralValidatorStatus = "active" | "pending" | "exited" | "withdrawal";
```
Comment (nflaig) [inline packages/beacon-node/src/api/impl/beacon/state/utils.ts]: no need to export this type

```suggestion
type GeneralValidatorStatus = "active" | "pending" | "exited" | "withdrawal";
```

**Review by nflaig — COMMENTED** [review]
LGTM - thanks for the cleanup

**Review by nflaig — APPROVED** [review]

**Review by wemeetagain — APPROVED** [review]
LGTM

unrelated, commenting for posterity
 `ValidatorStatus` shouldn't contain `"active"`

**Comment by nflaig** [comment]
> unrelated, commenting for posterity `ValidatorStatus` shouldn't contain `"active"`

yeah, let's get rid of this https://github.com/ChainSafe/lodestar/pull/7146

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.23.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.23.0) :tada:
